altos-rust/altos-core/libs/heap/free_list_allocator/src/free_list.rs

```diff
@@ -0,0 +1,494 @@
+/*
+ * Copyright (C) 2017  AltOS-Rust Team
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+//! Linked list code for the the memory allocator
+//! This is intended for use by the free_list_allocator
+//!
+
+use core::{mem, ptr};
+
+use alignment;
+
+// BlockHeader keeps track of a free block of memory
+#[cfg(not(test))]
+#[repr(C)]
```

Comment:
The thing I did with the tests is kind of funky, and I wasn't sure if it was a good way to do it. Basically, I created two versions of the same struct, but one of them had public fields for testing. The reason I did this was because there's some functions that seemed to make more sense in test.rs, but they can't use the fields if they're not public.